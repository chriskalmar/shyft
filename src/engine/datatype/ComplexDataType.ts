export class ComplexDataType {}

export const isComplexDataType = (obj: unknown): obj is ComplexDataType => {
  return obj instanceof ComplexDataType;
};
