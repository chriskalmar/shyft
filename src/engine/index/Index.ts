import { uniq } from 'lodash';
import { passOrThrow, isArray } from '../util';

import { Entity } from '../entity/Entity';

export const INDEX_UNIQUE = 'unique';
export const INDEX_GENERIC = 'generic';
export const indexTypes = [INDEX_UNIQUE, INDEX_GENERIC];

export type IndexSetup = {
  type?: string;
  attributes?: string[];
};

export class Index {
  type: string;
  attributes: string[];

  constructor(setup: IndexSetup = {} as IndexSetup) {
    const { type, attributes } = setup;

    passOrThrow(type, () => 'Missing index type');
    passOrThrow(
      indexTypes.indexOf(type) >= 0,
      () =>
        `Unknown index type '${type}' used, try one of these: '${indexTypes.join(
          ', ',
        )}'`,
    );

    passOrThrow(
      isArray(attributes, true),
      () =>
        `Index definition of type '${type}' needs to have a list of attributes`,
    );

    attributes.map((attribute) => {
      passOrThrow(
        typeof attribute === 'string',
        () =>
          `Index definition of type '${type}' needs to have a list of attribute names`,
      );
    });

    passOrThrow(
      attributes.length === uniq(attributes).length,
      () =>
        `Index definition of type '${type}' needs to have a list of unique attribute names`,
    );

    this.type = type;
    this.attributes = attributes;
  }

  toString() {
    return this.type;
  }
}

export const isIndex = (obj: unknown): obj is Index => {
  return obj instanceof Index;
};

export const processEntityIndexes = (entity: Entity, indexes: Index[]) => {
  passOrThrow(
    isArray(indexes),
    () =>
      `Entity '${entity.name}' indexes definition needs to be an array of indexes`,
  );

  const singleAttributeIndexes = [];

  const entityAttributes = entity.getAttributes();

  indexes.map((index, idx) => {
    passOrThrow(
      isIndex(index),
      () =>
        `Invalid index definition for entity '${entity.name}' at position '${idx}'`,
    );

    index.attributes.map((attributeName) => {
      passOrThrow(
        entityAttributes[attributeName],
        () =>
          `Cannot use attribute '${entity.name}.${attributeName}' in index as it does not exist`,
      );

      if (index.type === INDEX_UNIQUE) {
        passOrThrow(
          entityAttributes[attributeName].required,
          () =>
            `Cannot use attribute '${entity.name}.${attributeName}' in uniqueness index as it is nullable`,
        );

        passOrThrow(
          !entityAttributes[attributeName].i18n,
          () =>
            `Cannot use attribute '${entity.name}.${attributeName}' in uniqueness index as it is translatable`,
        );

        if (index.attributes.length === 1) {
          entityAttributes[attributeName].unique = true;
        }
      }

      if (index.attributes.length === 1) {
        singleAttributeIndexes.push(attributeName);
      }
    });
  });

  // add new index for single attributes with the 'index' flag set but not defined in an index object by the user
  Object.keys(entityAttributes).map((attributeName) => {
    if (entityAttributes[attributeName].index) {
      if (!singleAttributeIndexes.includes(attributeName)) {
        indexes.push(
          new Index({
            type: INDEX_GENERIC,
            attributes: [attributeName],
          }),
        );
      }
    }
  });

  return indexes;
};
