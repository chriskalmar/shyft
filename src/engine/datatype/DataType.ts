import { Entity, ShadowEntity, ViewEntity } from '../..';
import { passOrThrow, isFunction } from '../util';
import { ComplexDataType } from './ComplexDataType';

export type DataTypeSetup = {
  name?: string;
  description?: string;
  mock?: () => any;
  validate?: () => any;
  enforceRequired?: boolean;
  defaultValue?: () => any;
  enforceIndex?: boolean;
};

export type DataTypeFunction = (params: {
  setup: DataTypeSetup;
  entity?: Entity | ShadowEntity | ViewEntity;
}) => DataType | ComplexDataType;

export class DataType {
  name: string;
  description: string;
  mock?: () => any;
  validator?: (value: any, context: any) => any;
  enforceRequired?: boolean;
  defaultValue?: () => any;
  enforceIndex?: boolean;

  constructor(setup: DataTypeSetup = {} as DataTypeSetup) {
    const {
      name,
      description,
      mock,
      validate,
      enforceRequired,
      defaultValue,
      enforceIndex,
    } = setup;

    passOrThrow(name, () => 'Missing data type name');
    passOrThrow(
      description,
      () => `Missing description for data type '${name}'`,
    );

    passOrThrow(
      isFunction(mock),
      () => `'Missing mock function for data type '${name}'`,
    );

    if (validate) {
      passOrThrow(
        isFunction(validate),
        () => `'Invalid validate function for data type '${name}'`,
      );

      this.validator = validate;
    }

    if (defaultValue) {
      passOrThrow(
        isFunction(defaultValue),
        () => `'Invalid defaultValue function for data type '${name}'`,
      );

      this.defaultValue = defaultValue;
    }

    this.name = name;
    this.description = description;
    this.mock = mock;

    if (enforceRequired) {
      this.enforceRequired = enforceRequired;
    }

    if (enforceIndex) {
      this.enforceIndex = enforceIndex;
    }
  }

  validate = async (value: any, context: any): Promise<void> => {
    if (value && this.validator) {
      await this.validator(value, context);
    }
  };

  toString(): string {
    return this.name;
  }
}

export const isDataType = (obj: unknown): obj is DataType => {
  return obj instanceof DataType;
};
