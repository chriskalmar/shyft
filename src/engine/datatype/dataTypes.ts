import { DataType } from './DataType';
import { DataTypeUser } from './DataTypeUser';
import { randomJson } from '../util';
import * as casual from 'casual';

export const DataTypeUserID = new DataTypeUser({
  name: 'DataTypeUserID',
  description: 'Data type representing a reference to a user',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeID = new DataType({
  name: 'DataTypeID',
  description: 'Data type representing unique IDs',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeInteger = new DataType({
  name: 'DataTypeInteger',
  description: 'Data type representing integer values',
  /* istanbul ignore next */
  mock: () => casual.integer(-2 ^ 10, 2 ^ 10),
});

export const DataTypeBigInt = new DataType({
  name: 'DataTypeBigInt',
  description: 'Data type representing big integer values',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeFloat = new DataType({
  name: 'DataTypeFloat',
  description: 'Data type representing float values',
  /* istanbul ignore next */
  mock: () => casual.double(-2 ^ 10, 2 ^ 10),
});

export const DataTypeDouble = new DataType({
  name: 'DataTypeDouble',
  description: 'Data type representing double precision values',
  /* istanbul ignore next */
  mock: () => casual.double(-2 ^ 10, 2 ^ 10),
});

export const DataTypeBoolean = new DataType({
  name: 'DataTypeBoolean',
  description: 'Data type representing boolean values',
  /* istanbul ignore next */
  mock: () => casual.boolean,
  enforceRequired: true,
  defaultValue() {
    return false;
  },
});

export const DataTypeString = new DataType({
  name: 'DataTypeString',
  description: 'Data type representing text values',
  /* istanbul ignore next */
  mock: () => casual.title,
});

export const DataTypeJson = new DataType({
  name: 'DataTypeJson',
  description: 'Data type representing json objects',
  /* istanbul ignore next */
  mock: randomJson,
});

export const DataTypeTimestamp = new DataType({
  name: 'DataTypeTimestamp',
  description: 'Data type representing a timestamp',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTimestampTz = new DataType({
  name: 'DataTypeTimestampTz',
  description: 'Data type representing a timestamp with time zone information',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeDate = new DataType({
  name: 'DataTypeDate',
  description: 'Data type representing a date',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTime = new DataType({
  name: 'DataTypeTime',
  description: 'Data type representing time',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTimeTz = new DataType({
  name: 'DataTypeTimeTz',
  description: 'Data type representing time with time zone information',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeUUID = new DataType({
  name: 'DataTypeUUID',
  description: 'Data type representing a UUID',
  /* istanbul ignore next */
  mock: () => casual.uuid,
});

export const DataTypeI18n = new DataType({
  name: 'DataTypeI18n',
  description: 'Data type representing translation objects',
  /* istanbul ignore next */
  mock: randomJson,
});

export const DataTypeUpload = new DataType({
  name: 'DataTypeUpload',
  description: 'Data type representing a file upload',
  mock: () => null,
});
