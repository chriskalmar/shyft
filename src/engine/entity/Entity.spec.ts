/* eslint-disable @typescript-eslint/no-explicit-any */
import { graphql, GraphQLSchema } from 'graphql';
import { Entity, isEntity } from './Entity';
import { Index, INDEX_UNIQUE, isIndex } from '../index/Index';
import {
  Mutation,
  isMutation,
  MUTATION_TYPE_CREATE,
} from '../mutation/Mutation';
import { Permission, isPermission } from '../permission/Permission';
import { passOrThrow } from '../util';
import { DataTypeID, DataTypeString } from '../datatype/dataTypes';

import { generateGraphQLSchema } from '../../graphqlProtocol/generator';
import { generateTestSchema } from '../../graphqlProtocol/test-helper';

describe('Entity', () => {
  it('should have a name', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Entity({} as any);
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have a description', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Entity({
        name: 'Example',
      } as any);
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have a map of attributes', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Entity({
        name: 'Example',
        description: 'Just some description',
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it("should return it's name", () => {
    const entity = new Entity({
      name: 'SomeEntityName',
      description: 'Just some description',
      attributesGenerator: () => {
        return {} as any;
      },
    });

    expect(entity.name).toBe('SomeEntityName');
    expect(String(entity)).toBe('SomeEntityName');
  });

  it('should accept only maps or functions as attributes definition', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Entity({
        name: 'Example',
        description: 'Just some description',
        attributes: [2, 7, 13] as any,
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should reject non-map results of attribute definition functions', () => {
    function fn() {
      // eslint-disable-next-line no-new
      const entity = new Entity({
        name: 'Example',
        description: 'Just some description',
        attributesGenerator: () => {
          return [2, 7, 13] as any;
        },
      });

      entity.getAttributes();
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  describe('should have a list of attributes', () => {
    it('as a map', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          id: {
            type: DataTypeID,
            description: 'ID of item',
            primary: true,
          },
          name: {
            type: DataTypeString,
            description: 'Just another description',
          },
        },
      });

      const attributes = entity.getAttributes();

      expect(attributes.id.type).toBe(DataTypeID);
      expect(attributes.name.type).toBe(DataTypeString);
    });

    it('as a function return a map', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributesGenerator: () => {
          return {
            id: {
              type: DataTypeID,
              description: 'ID of item',
              primary: true,
            },
            name: {
              type: DataTypeString,
              description: 'Just another description',
            },
          };
        },
      });

      const attributes = entity.getAttributes();

      expect(attributes.id.type).toBe(DataTypeID);
      expect(attributes.name.type).toBe(DataTypeString);
    });
  });

  it('should throw if invalid storage type was provided', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new Entity({
        name: 'Example',
        description: 'Just some description',
        attributes: {
          id: {
            type: DataTypeID,
            description: 'ID of item',
            primary: true,
          },
        },
        storageType: {},
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  describe('isEntity', () => {
    it('should recognize objects of type Entity', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributesGenerator() {
          return null as any;
        },
      });

      function fn() {
        passOrThrow(isEntity(entity), () => 'This error will never happen');
      }

      expect(fn).not.toThrow();
    });

    it('should recognize non-Entity objects', () => {
      function fn() {
        passOrThrow(
          // eslint-disable-next-line @typescript-eslint/no-empty-function
          isEntity({}) || isEntity(function test() {}) || isEntity(Error),
          () => 'Not an Entity object',
        );
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('attributes', () => {
    it('should catch invalid attribute names', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          ['wrong-named-attribute']: {
            type: DataTypeID,
            description: 'Something',
          },
        },
      });

      function fn() {
        entity.getAttributes();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should reject an empty attributes map', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {},
      });

      function fn() {
        entity.getAttributes();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should reject attributes without a description', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          someAttribute: {} as any,
        },
      });

      function fn() {
        entity.getAttributes();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should reject attributes with missing or invalid data type', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          someAttribute: {
            description: 'Just some description',
          } as any,
        },
      });

      function fn() {
        entity.getAttributes();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should accept attributes with valid data types or references to other entities', () => {
      const Country = new Entity({
        name: 'Country',
        description: 'A country',
        attributes: {
          name: {
            type: DataTypeString,
            description: 'Country name',
          },
        },
      });

      const City = new Entity({
        name: 'City',
        description: 'A city in a country',
        attributes: {
          name: {
            type: DataTypeString,
            description: 'City name',
          },
          country: {
            type: Country,
            description: 'Belongs to a country',
          },
        },
      });

      const attributes = City.getAttributes();

      expect(String(attributes.name.type)).toBe('DataTypeString');
      expect(String(attributes.country.type)).toBe('Country');
    });

    it('should throw if provided with an invalid resolve function', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          someAttribute: {
            type: DataTypeString,
            description: 'Just some description',
            resolve: {} as any,
          },
        },
      });

      function fn() {
        entity.getAttributes();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should throw if provided with an invalid defaultValue function', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          someAttribute: {
            type: DataTypeString,
            description: 'Just some description',
            defaultValue: 'not-a-function' as any,
          },
        },
      });

      function fn() {
        entity.getAttributes();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should throw if provided with an invalid validate function', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          someAttribute: {
            type: DataTypeString,
            description: 'Just some description',
            validate: 'not-a-function' as any,
          },
        },
      });

      function fn() {
        entity.getAttributes();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('primary attribute', () => {
    it('should catch multiple primary attributes', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          someAttribute: {
            type: DataTypeID,
            description: 'Something',
            primary: true,
          },
          anotherAttribute: {
            type: DataTypeID,
            description: 'Something else',
            primary: true,
          },
        },
      });

      function fn() {
        entity.getAttributes();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should catch primary attributes with invalid data types', () => {
      const Country = new Entity({
        name: 'Country',
        description: 'A country',
        attributes: {
          name: {
            type: DataTypeString,
            description: 'Country name',
          },
        },
      });

      const City = new Entity({
        name: 'City',
        description: 'A city',
        attributes: {
          someAttribute: {
            type: Country,
            description: 'Something',
            primary: true,
          },
        },
      });

      function fn() {
        City.getAttributes();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('references', () => {
    const Country = new Entity({
      name: 'Country',
      description: 'A country',
      attributes: {
        name: {
          type: DataTypeString,
          description: 'Country name',
        },
      },
    });

    it('should return a request attribute when used as reference', () => {
      const countryName = Country.referenceAttribute('name');

      expect(String(countryName.type)).toBe('DataTypeString');
    });

    it('should throw if invalid attribute is to be referenced', () => {
      function fn() {
        Country.referenceAttribute('notHere');
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('system attributes', () => {
    it('should extend model with time tracking attributes if requested', () => {
      const Country = new Entity({
        name: 'Country',
        description: 'A country',
        includeTimeTracking: true,
        attributes: {
          name: {
            type: DataTypeString,
            description: 'Country name',
          },
        },
      });

      const attributes = Country.getAttributes();

      expect(String(attributes.createdAt.type)).toBe('DataTypeTimestampTz');
      expect(String(attributes.updatedAt.type)).toBe('DataTypeTimestampTz');
    });

    it('should extend model with user tracking attributes if requested', () => {
      const Country = new Entity({
        name: 'Country',
        description: 'A country',
        includeUserTracking: true,
        attributes: {
          name: {
            type: DataTypeString,
            description: 'Country name',
          },
        },
      });

      const attributes = Country.getAttributes();

      expect(String(attributes.createdBy.type)).toBe('DataTypeUserID');
      expect(String(attributes.updatedBy.type)).toBe('DataTypeUserID');
    });

    it('should throw if user defined attribute name collides with a system attribute name', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        includeTimeTracking: true,
        attributes: {
          updatedAt: {
            type: DataTypeString,
            description: 'Just some description',
          },
        },
      });

      function fn() {
        entity.getAttributes();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('indexes', () => {
    it('should set unique flag on attributes based on index definition', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          loginName: {
            type: DataTypeString,
            description: 'Just some description',
            required: true,
          },
          firstName: {
            type: DataTypeString,
            description: 'Just some description',
            required: true,
          },
          lastName: {
            type: DataTypeString,
            description: 'Just some description',
            required: true,
          },
          email: {
            type: DataTypeString,
            description: 'Just some description',
            required: true,
          },
        },
        indexes: [
          new Index({
            type: INDEX_UNIQUE,
            attributes: ['loginName'],
          }),
          new Index({
            type: INDEX_UNIQUE,
            attributes: ['firstName', 'lastName'],
          }),
          new Index({
            type: INDEX_UNIQUE,
            attributes: ['email'],
          }),
        ],
      });

      entity.getIndexes();
      const attributes = entity.getAttributes();

      expect(attributes.loginName.unique).toBe(true);
      expect(attributes.firstName.unique).not.toBe(true);
      expect(attributes.lastName.unique).not.toBe(true);
      expect(attributes.email.unique).toBe(true);
    });

    it('should return a list of indexes', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          someAttribute: {
            type: DataTypeString,
            description: 'Just some description',
            required: true,
          },
        },
        indexes: [
          new Index({
            type: INDEX_UNIQUE,
            attributes: ['someAttribute'],
          }),
        ],
      });

      entity.getAttributes();
      const indexes = entity.getIndexes();

      expect(Array.isArray(indexes)).toBe(true);
      expect(isIndex(indexes[0])).toBe(true);
    });
  });

  describe('mutations', () => {
    it('should return a list of mutations', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          someAttribute: {
            type: DataTypeString,
            description: 'Just some description',
          },
        },
        mutations: [
          new Mutation({
            type: MUTATION_TYPE_CREATE,
            name: 'build',
            description: 'build item',
            attributes: ['someAttribute'],
          }),
        ],
      });

      const mutations = entity.getMutations();

      expect(Array.isArray(mutations)).toBe(true);
      expect(isMutation(mutations[0])).toBe(true);
    });

    it('should automatically have default mutations', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          someAttribute: {
            type: DataTypeString,
            description: 'Just some description',
          },
        },
      });

      entity.getAttributes();

      expect(isMutation(entity.getMutationByName('create'))).toBe(true);
      expect(isMutation(entity.getMutationByName('update'))).toBe(true);
      expect(isMutation(entity.getMutationByName('delete'))).toBe(true);
    });

    it('should allow to add default mutations', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          someAttribute: {
            type: DataTypeString,
            description: 'Just some description',
          },
        },
        mutations: ({ createMutation, deleteMutation }) => [
          createMutation,
          deleteMutation,
        ],
      });

      entity.getAttributes();

      expect(isMutation(entity.getMutationByName('create'))).toBe(true);
      expect(isMutation(entity.getMutationByName('update'))).toBe(false);
      expect(isMutation(entity.getMutationByName('delete'))).toBe(true);
    });
  });

  describe('permissions', () => {
    it('should return a list of permissions', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          someAttribute: {
            type: DataTypeString,
            description: 'Some description',
          },
        },
        permissions: {
          read: new Permission().authenticated(),
        },
      });

      const permissions = entity.getPermissions();

      expect(isPermission(permissions.read)).toBe(true);
    });

    it('should throw if empty permissions are provided', () => {
      const entity = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          someAttribute: {
            type: DataTypeString,
            description: 'Some description',
          },
        },
        permissions: {
          read: new Permission(),
          find: [new Permission().authenticated(), new Permission()],
          mutations: {
            delete: new Permission(),
          },
        },
      });

      function fn() {
        entity.getPermissions();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('states', () => {
    const entityDefinition = {
      name: 'SomeEntityName',
      description: 'Just some description',
      attributes: {
        something: {
          type: DataTypeString,
          description: 'Just some description',
        },
      },
    };

    const states = {
      open: 10,
      closed: 20,
      inTransfer: 40,
      onHold: 50,
    };

    it('should return a list of states', () => {
      const entity = new Entity({
        ...entityDefinition,
        states,
      });

      const theStates = entity.getStates();

      expect(states).toEqual(theStates);
    });

    it('should throw if provided with an invalid map of states', () => {
      function fn() {
        const entity = new Entity({
          ...entityDefinition,
          states: ['bad'],
        });

        entity.getStates();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should throw on invalid state names', () => {
      function fn() {
        const entity = new Entity({
          ...entityDefinition,
          states: {
            ['bad-state-name']: 123,
          },
        });

        entity.getStates();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should throw on invalid state IDs', () => {
      function fn1() {
        const entity = new Entity({
          ...entityDefinition,
          states: {
            open: 1.234,
          },
        });

        entity.getStates();
      }

      expect(fn1).toThrowErrorMatchingSnapshot();

      function fn2() {
        const entity = new Entity({
          ...entityDefinition,
          states: {
            open: -1,
          },
        });

        entity.getStates();
      }

      expect(fn2).toThrowErrorMatchingSnapshot();

      function fn3() {
        const entity = new Entity({
          ...entityDefinition,
          states: {
            open: 0,
          },
        });

        entity.getStates();
      }

      expect(fn3).toThrowErrorMatchingSnapshot();

      function fn4() {
        const entity = new Entity({
          ...entityDefinition,
          states: {
            open: 'not a number',
          },
        });

        entity.getStates();
      }

      expect(fn4).toThrowErrorMatchingSnapshot();
    });

    it('should throw if state IDs are not unique', () => {
      function fn() {
        const entity = new Entity({
          ...entityDefinition,
          states: {
            open: 100,
            closed: 100,
          },
        });

        entity.getStates();
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });

  describe('preProcessor', () => {
    let graphqlSchema: GraphQLSchema;

    beforeAll(async () => {
      const preProcessor = (entity, source, args, context, info) => {
        ({ context, info }); // overcome linter warnings

        return args;
      };

      const postProcessor = (
        result,
        entity,
        id,
        source,
        input,
        typeName,
        mutation,
        context,
      ) => {
        ({ entity, id, source, input, typeName, mutation, context }); // overcome linter warnings

        result.something = 'someotherthing';
        return result;
      };

      const SomeEntityWithPreprocess = new Entity({
        name: 'SomeEntityName',
        description: 'Just some description',
        attributes: {
          id: {
            type: DataTypeID,
            description: 'just some id',
            primary: true,
          },
          something: {
            type: DataTypeString,
            description: 'Just some description',
          },
        },
        preProcessor,
      });

      const SomeEntityWithPostprocess = new Entity({
        name: 'SomeOtherEntityName',
        description: 'Just some other description',
        attributes: {
          id: {
            type: DataTypeID,
            description: 'just some id',
            primary: true,
          },
          something: {
            type: DataTypeString,
            description: 'Just some description',
          },
        },
        postProcessor,
      });

      // create mutations to test processors ?
      const setup = await generateTestSchema([
        SomeEntityWithPreprocess,
        SomeEntityWithPostprocess,
      ]);
      const configuration = setup.configuration;
      graphqlSchema = generateGraphQLSchema(configuration);
    });

    it('should have a valid preProcessor function', () => {
      function fn() {
        // eslint-disable-next-line no-new
        new Entity({
          name: 'SomeEntityName',
          description: 'Just some description',
          attributes: {
            something: {
              type: DataTypeString,
              description: 'Just some description',
            },
          },
          preProcessor: 'not-a-function',
        });
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should have a valid postProcessor function', () => {
      function fn() {
        // eslint-disable-next-line no-new
        new Entity({
          name: 'SomeEntityName',
          description: 'Just some description',
          attributes: {
            something: {
              type: DataTypeString,
              description: 'Just some description',
            },
          },
          postProcessor: 'not-a-function',
        });
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });

    it('should pass through preProcessor if it is declared', async () => {
      const query = `
        query AllSomeEntityNames {
          allSomeEntityNames {
            edges {
              node {
                id
                something
              }
            }
          }
        }`;

      const result = await graphql(
        graphqlSchema,
        query,
        null,
        { userId: 10 },
        // { filter: {} },
      );

      expect(result).toMatchSnapshot();
    });

    it('should pass through postProcessor if it is declared', async () => {
      const query = `
        query AllSomeOtherEntityNames {
          allSomeOtherEntityNames {
            edges {
              node {
                id
                something
              }
            }
          }
        }`;

      const result = await graphql(
        graphqlSchema,
        query,
        null,
        {},
        // { filter: {} },
      );

      expect(result).toMatchSnapshot();
    });
  });
});
