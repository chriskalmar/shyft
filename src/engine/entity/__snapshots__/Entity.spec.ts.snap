// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity attributes should catch invalid attribute names 1`] = `"Invalid attribute name 'wrong-named-attribute' in entity 'SomeEntityName' (Regex: /^[a-zA-Z][a-zA-Z0-9_]*$/)"`;

exports[`Entity attributes should reject an empty attributes map 1`] = `"Entity 'SomeEntityName' has no attributes defined"`;

exports[`Entity attributes should reject attributes with missing or invalid data type 1`] = `"'SomeEntityName.someAttribute' has invalid data type 'undefined'"`;

exports[`Entity attributes should reject attributes without a description 1`] = `"Missing description for 'SomeEntityName.someAttribute'"`;

exports[`Entity attributes should throw if provided with an invalid defaultValue function 1`] = `"'SomeEntityName.someAttribute' has an invalid defaultValue function'"`;

exports[`Entity attributes should throw if provided with an invalid resolve function 1`] = `"'SomeEntityName.someAttribute' has an invalid resolve function'"`;

exports[`Entity attributes should throw if provided with an invalid validate function 1`] = `"'SomeEntityName.someAttribute' has an invalid validate function'"`;

exports[`Entity isEntity should recognize non-Entity objects 1`] = `"Not an Entity object"`;

exports[`Entity permissions should throw if empty permissions are provided 1`] = `"Entity 'SomeEntityName' has one or more empty permission definitions in: read, find, mutations.delete"`;

exports[`Entity preProcessor should have a valid postProcessor function 1`] = `"postProcessor of entity 'SomeEntityName' needs to be a valid function"`;

exports[`Entity preProcessor should have a valid preProcessor function 1`] = `"preProcessor of entity 'SomeEntityName' needs to be a valid function"`;

exports[`Entity preProcessor should pass through postProcessor if it is declared 1`] = `
Object {
  "data": Object {
    "allSomeOtherEntityNames": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "1",
            "something": "someotherthing",
          },
        },
      ],
    },
  },
}
`;

exports[`Entity preProcessor should pass through preProcessor if it is declared 1`] = `
Object {
  "data": Object {
    "allSomeEntityNames": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "1",
            "something": "something",
          },
        },
      ],
    },
  },
}
`;

exports[`Entity primary attribute should catch multiple primary attributes 1`] = `"'SomeEntityName.anotherAttribute' cannot be set as primary attribute,'someAttribute' is already the primary attribute"`;

exports[`Entity primary attribute should catch primary attributes with invalid data types 1`] = `"Primary attribute 'City.someAttribute' has invalid data type 'Country'"`;

exports[`Entity references should throw if invalid attribute is to be referenced 1`] = `"Cannot reference attribute 'Country.notHere' as it does not exist"`;

exports[`Entity should accept only maps or functions as attributes definition 1`] = `"'attributes' for entity 'Example' needs to be a map of attributes or a function returning a map of attributes"`;

exports[`Entity should have a description 1`] = `"Missing description for entity 'Example'"`;

exports[`Entity should have a map of attributes 1`] = `"'attributes' for entity 'Example' needs to be a map of attributes or a function returning a map of attributes"`;

exports[`Entity should have a name 1`] = `"Missing entity name"`;

exports[`Entity should reject non-map results of attribute definition functions 1`] = `"Attribute definition function for entity 'Example' does not return a map"`;

exports[`Entity should throw if invalid storage type was provided 1`] = `"Entity 'Example' needs a valid storage type (defaults to 'StorageTypeNull')"`;

exports[`Entity states should throw if provided with an invalid map of states 1`] = `"Entity 'SomeEntityName' states definition needs to be a map of state names and their unique ID"`;

exports[`Entity states should throw if state IDs are not unique 1`] = `"Each state defined in entity 'SomeEntityName' needs to have a unique ID"`;

exports[`Entity states should throw on invalid state IDs 1`] = `"State 'open' in entity 'SomeEntityName' has an invalid unique ID (needs to be a positive integer)"`;

exports[`Entity states should throw on invalid state IDs 2`] = `"State 'open' in entity 'SomeEntityName' has an invalid unique ID (needs to be a positive integer)"`;

exports[`Entity states should throw on invalid state IDs 3`] = `"State 'open' in entity 'SomeEntityName' has an invalid unique ID (needs to be a positive integer)"`;

exports[`Entity states should throw on invalid state IDs 4`] = `"State 'open' in entity 'SomeEntityName' has an invalid unique ID (needs to be a positive integer)"`;

exports[`Entity states should throw on invalid state names 1`] = `"Invalid state name 'bad-state-name' in entity 'SomeEntityName' (Regex: /^[a-zA-Z][_a-zA-Z0-9]*$/)"`;

exports[`Entity system attributes should throw if user defined attribute name collides with a system attribute name 1`] = `"Attribute name collision with system attribute 'updatedAt' in entity 'SomeEntityName'"`;
